### ISSUE TITLE
Best practice for delaying say() function? 

### ISSUE THREAD

**Issue by sheafallick** [slackapi/bolt-python#795]
Hello! I was wondering if anyone in this community has tried delaying the say() function for days after an event occurs. If so, would you mind sharing your code or any takeaways from the experience? 

My project is a drip campaign that's sparked from a "team_join" event. I want a DM to go out to the Slack member immediately (already implemented this) and then send a DM every few days with follow-up information (not implemented). I'm thinking of using time.sleep() or event.wait(), but would love to hear from other folks. 

Thank you!

**Comment by WilliamBergamin** [comment]
Hi @sheafallick thanks for writing in!

This is a good question for this use case, there are a few ways I can think of to implement this.

One that stands out, if you know exactly what the messages should be, takes advantage of [scheduledMessages](https://api.slack.com/methods/chat.scheduleMessage). In short once the app receives an event it creates a series of `ScheduledMessages` then uses the `say()` function to reply to the user.

The code could look something like this 

```python
from logging import Logger

from slack_bolt import BoltContext, Say
from slack_sdk import WebClient


def sample_message_callback(context: BoltContext, say: Say, client: WebClient, logger: Logger):
    try:
        client.chat_scheduleMessage(
            channel = "C1234567890",
            post_at = "299876400",
            text = "hello in the future"
        )
        greeting = context["matches"][0]
        say(f"{greeting}, how are you?")
    except Exception as e:
        logger.error(e)
``` 

[Here is more info on `scheduledMessages`](https://api.slack.com/messaging/scheduling) Let me know if this is helpful!

**Comment by sheafallick** [comment]
Ah amazing, thank you @WilliamBergamin! For the "post_at" field, could I make it dynamic by saving the timestamp of the event_join and adding a set amount of time after (i.e. team_join timestamp for user x + 5 days)? We don't have a database currently so I'm particularly interested in solutions where we can save those variables locally. 

If you don't know, I can do some more research here. Thank you for your help!

**Comment by WilliamBergamin** [comment]
Absolutely, you could fetch the event time from the payload and add 5 days with something like this 
```python
def sample_message_callback(event, context: BoltContext, say: Say, client: WebClient, logger: Logger):
    try:
        next_message_time = event["event_time"] + 432000
        client.chat_scheduleMessage(
            channel = context.channel_id,
            post_at =next_message_time,
            text = "hello in the future"
        )
    except Exception as e:
        logger.error(e)
```

I would also tend to stay away from a database, since it introduces complexity, instead it is possible to perform the following operations on `scheduledMessages`

- [create](https://api.slack.com/methods/chat.scheduleMessage)
- [list](https://api.slack.com/methods/chat.scheduledMessages.list)
- [delete](https://api.slack.com/methods/chat.deleteScheduledMessage)

**Comment by sheafallick** [comment]
Amazing, thank you so much @WilliamBergamin! I'm going to build this next week assuming Marketing gets the copy to me, I'll follow up and let you know how the effort goes.

**Comment by sheafallick** [comment]
@WilliamBergamin, thanks so much for your help! I was able to go off your advice & figure it out from there. A couple of notes from my implementation:

1. I needed to convert the ts in the event data to a float and then round it to the nearest second to get the time logic to work (the input for the `client.chat_scheduleMessage` post_at parameter needs to be an integer). 
2. Using the "say" function w/ the greeting variable broke the code for me. I found just writing the `client.chat_scheduleMessage` sections of the code for each message within the function associated with the listener worked fine.

I'll close this now, I think it's working. If there are any bugs I encounter that I can't figure out I'd love to pick your brain about them. Thanks again for your time 😄
